### ISSUE TITLE
Bridge.define is not helping with Intellisense in Visual Studio

### ISSUE THREAD

**Issue by artfuldev** [bridgedotnet/Bridge#249]
Since there are `Bridge.define` statements instead of normal object/variable initializers in the javascript generated, Intellisense suffers. If possible, we should improve upon that factor also.

One of the reasons I came to use Bridge.NET is because I wanted my JS to stay in sync with my back-end C# code. Keeping in sync also means helping the JavaScript developers with the Intellisense so that they make less typos and don't get lost so often (I'm working on a medium-to-large-sized project).

**Comment by fabriciomurta** [comment]
Hello, we are trying to make the most out of intellisense and documentation as to remove any need of `dynamic()` or hand-written `Script.Write()`.

But in the case, you want intellisense working on the generated JavaScript code? I don't quite get how exactly intellisense is suffering. Can you show an example or something like it?

**Comment by kenshinthebattosai** [comment]
Yes, I am talking about intellisense from the generated javascript.

This is my case:
We are working on a really large project.
I have a class like this:

```c#
using System;

namespace Helpers
{
    public static class ViewTypes
    {
        public const String ListView = "listview";
        public const String TitleView = "titleview";
        public const String GridView = "gridview";
        public const String MapView = "mapview";
    }
}
```

This is used across to get and set the view type, say, across the cross-platform C# application.
It would be wonderful if we could do the same in javascript. When a C# developer of my team uses Visual Studio to write code, he can at once see the different ViewTypes allowed by typing `Helpers.ViewTypes.` in the IDE and the Intellisense would pop up which will minimize typos etc, I hope you get the point. I wanted the same for the JavaScript generated. I am using Bridge.NET, basically, to maintain sync between the C# and JS code.

This is the JS code that Bridge.NET generates:
```javascript
Bridge.define('Helpers.ViewTypes', {
    statics: {
        ListView: "listview",
        TitleView: "titleview",
        GridView: "gridview",
        MapView: "mapview"
    }
});
```

The Intellisense works in the console, most definitely. I have tried Google Chrome and when I tried the same in the Developer Console accessed by pressing F12 and typing `Helpers.ViewTypes.`. However, when a javascript developer is writing code in Visual Studio, he has no idea of even the existence of the Helpers namespace (JS object) so does not get Intellisense. This would be available, however, if the generated code was something along the lines of this:

```javascript
if (typeof(Helpers) == "undefined")
    var Helpers = {};
Helpers.ViewTypes = function (){
}; //{}
Helpers.ViewTypes.ListView = "listview";
Helpers.ViewTypes.TitleView = "titleview";
Helpers.ViewTypes.GridView = "gridview";
Helpers.ViewTypes.MapView = "mapview";
```

**Comment by fabriciomurta** [comment]
Thank you very much for the clear and comprehensive explanation, this sounds like a fantastic feature indeed but maybe will take some time to be developed. We'll update here as we progress towards this.

**Comment by fabriciomurta** [comment]
This would be the base for developing this functionality but, as it does not really looks trivial, should take some time to get done.
https://msdn.microsoft.com/en-us/library/bb385682.aspx#Script

**Comment by kenshinthebattosai** [comment]
I would be very much willing to put in some work in this area, if that is alright. I am yet to have a thorough look at the source code so if you can give me some pointers it would be helpful.

**Comment by fabriciomurta** [comment]
Sure, that'll be great! But to avoid problems, we've come up with [a CLA you should agree](https://github.com/bridgedotnet/Bridge/blob/master/CLA.md) to before we can accept your valuable additions.

For that diplomacy, please review [CONTRIBUTING.md](https://github.com/bridgedotnet/Bridge/blob/master/CONTRIBUTING.md), then complete the following [Contributor Form](https://docs.google.com/a/object.net/forms/d/12IAxRmWCk2vD-M3mj4d0ihztqKoiREpRRSVq4pEk0-k/viewform?c=0&w=1).

Basically it says you're releasing your code under the same Apache 2.0 License Agreement.

**Comment by fabriciomurta** [comment]
As for the part of code you should work it will be probably on the `Translator.cs` (bridgedotnet/Builder repo, directory `Builder/Translator/Translator/Translator.cs`).

Most of the changes seem to be done there (on the `Bridge.Translator` namespace), so that's really the starting point.
As to, for example, add the `/// reference` lines it is the `saveTo()` function you want. You'll see there a command to write `/* global Bridge */` to the first line of every non-minified .js file generated, for example. We wouldn't need to bother with intellisense on the minified files, would we?

But the catch here might be building the references' file list, which would need some investigation to come up with the correct syntax. You'll have to query all the types that are created. A loop that might give you the result is in the `Bridge.Translator.Translator.InspectTypes()`, or `Bridge.Translator.Translator.BuildSyntaxTree()`. 

I think that, as pointers out of thought is what I can think so far. Further details requires investigation.

**Comment by kenshinthebattosai** [comment]
I tried cloning the `bridgedotnet/Builder` repo but it seems there are some problems with the location of the references. Seems there must be a parent Bridge directory? I have fixed the references and migrated to automatic package restore with a local change but I guess it cannot be submitted as a pull request? Can you please [have a look](https://github.com/kenshinthebattosai/Builder/commit/1297a5bef6b7b2a24874f86b72c12e79aabe2a6a) and share the actual folder structure required to work on this project, if there is something like that?

**Comment by geoffreymcgill** [comment]
A parent folder is required for the .sln files to reference the projects correctly. I don't think that folder needs to be named `Bridge`, but that is what I have set and it works.

All the Bridge related repo's should be cloned to this this `Bridge` folder. 

You can make it work if the repo's are cloned to various locations, but the .sln references would have to be revised. 

Hope this helps.

**Comment by fabriciomurta** [comment]
I believe issue #269 being implemented should avoid this kind of confusion in the future, ensuring the folders are in the right place so solution links are resolved successfully.

**Comment by kenshinthebattosai** [comment]
Yes, thank you. :)

**Comment by michaelcheers** [comment]
@artfuldev, Can't you use typescript definitions?

**Comment by artfuldev** [comment]
@michaelcheers this was reported a year ago, my friend...

**Comment by geoffreymcgill** [comment]
As mentioned by @michaelcheers, the recommended approach is to use the TypeScript **.d.ts** files generated by Bridge to accurately provide Intellisense within pure JavaScript apps.
